libraries/AP_Networking/AP_Networking_PPP.cpp

```diff
@@ -40,36 +44,101 @@ extern const AP_HAL::HAL& hal;
 #define PPP_LINK_TIMEOUT_MS 15000U
 #endif
 
+// each PPP frame has a 0x7E at start and end, which needs
+// to be stripped for network capture
+#define PPP_FRAME_BYTE 0x7E
+#define PPP_ESCAPE_BYTE 0x7D
+
 /*
   output some data to the uart
  */
 uint32_t AP_Networking_PPP::ppp_output_cb(ppp_pcb *pcb, const void *data, uint32_t len, void *ctx)
 {
     auto &driver = *(AP_Networking_PPP::PPP_Instance *)ctx;
     LWIP_UNUSED_ARG(pcb);
-    uint32_t remaining = len;
     const uint8_t *ptr = (const uint8_t *)data;
 
     if (pcb->phase == PPP_PHASE_TERMINATE) {
         // don't output while terminating
         return 0;
     }
 
-    if (driver.uart->txspace() < remaining) {
+    if (driver.uart->txspace() < len) {
         /*
           if we can't send the whole frame then don't send any of it. This
           minimises issues with the PPP state machine
          */
         return 0;
     }
 
-    return driver.uart->write(ptr, remaining);
+
+#if AP_NETWORKING_CAPTURE_ENABLED
+    /*
+      capture outgoing data
+     */
+    if (driver.backend->frontend.option_is_set(AP_Networking::OPTION::CAPTURE_PACKETS)) {
+        driver.capture_data(ptr, len);
+    }
+#endif
+
+    return driver.uart->write(ptr, len);
+}
+
+#if AP_NETWORKING_CAPTURE_ENABLED
+/*
+  capture a outgoing data packet
+ */
+void AP_Networking_PPP::PPP_Instance::capture_data(const uint8_t *ptr, uint32_t len)
+{
+    WITH_SEMAPHORE(capture.sem);
+    if (capture.fd == -1) {
+        return;
+    }
+    if (len >= 2 && ptr[0] == PPP_FRAME_BYTE) {
+        // strip framing flags
+        ptr++;
+        len -= 2;
+    }
+    /*
+      for PPP outgoing we need to unescape the data, but we
+      need the length for the header, which means we need to
+      walk the data twice
+    */
+    uint32_t dlen = 0;
+    for (uint32_t i = 0; i < len; i++) {
+        if (ptr[i] == PPP_ESCAPE_BYTE) {
+            i++;
+        }
+        dlen++;
+    }
+    auto &fs = AP::FS();
+    backend->capture_header(capture.fd, dlen+1); // +1 for direction byte
+    const uint8_t direction = 0; // direction out
+    fs.write(capture.fd, &direction, 1);
+
+    uint8_t buf[32];
+    uint32_t buflen = 0;
+    for (uint32_t i = 0; i < len; i++) {
+        if (ptr[i] == PPP_ESCAPE_BYTE && i + 1 < len) {
```

Comment:
What does happen if the last byte needs escaping?

Reply:
it sends the byte (which apparently is the right thing to do)